Python 3: Fix error with accPropServer

feerrenrut opened the pull request:
### Link to issue number:
#9768

### Summary of the issue:
The `VARIANT` out-param had not been initialised correct.
Comtypes was not initialising `VARIANT.vt` to `VT_EMPTY`, and attempting to call `VariantClear` (oleauto.h) before assigning it the value returned from `getProp`.

### Description of how this pull request fixes the issue:
Instead, we implement a low-level (require a `this` param) `getProp` method, which is then provided pointers to the out-params, and can initialise them correctly.

### Testing performed:
Run locally from source.

### Known issues with pull request:
None

### Change log entry:
None

feerrenrut committed:
Fix error with accPropServer

The VARIANT out-param had not been initialised correct.
Comtypes was not initialising VARIANT.vt to VT_EMPTY, and attempting to call VariantClear (oleauto.h) before assigning it the value returned from getProp.
Instead, we implement a low-level (require a this param) getProp method, which is then provided pointers to the out-params, and can initialise them correctly.

feerrenrut commented:
I'll attach a try build link shortly.

feerrenrut commented:
Please use the following [try build](https://ci.appveyor.com/api/buildjobs/7g7yv1k4u9xlikjx/artifacts/output%2Fnvda_snapshot_try-fixErrorInSettingDialog-17996%2C8d818031.exe) to test for issues.

leonardder commented on source/gui/accPropServer.py:
```diff
@@ -35,10 +37,8 @@ class IAccPropServer_Impl(COMObject, metaclass=ABCMeta):
-	# Consider using `NO_RETURN_VALUE`
-	NO_RETURN_VALUE = (VT_EMPTY, DOES_NOT_HAVE_PROP)
+
+	S_OK = 0
```
I think you both imported this above and defined this here.

leonardder commented on source/gui/accPropServer.py:
```diff
@@ -90,24 +91,53 @@ def _getPropValue(self, pIDString, dwIDStringLen, idProp):
+		see L{_getPropValue} for more details of args.
+		Uses a low-level approach, because comtypes tries to clear the VARIANT even though it is an out param.
+		When the pfHasProp part is FALSE / self.DOES_NOT_HAVE_PROP, then the pvarValue.vt part must be VT_EMPTY.
+		@type pvarValue: POINTER[VARIANT]
```
Don't type hints work here? Something like POINTER(variant) should just create a new class.

feerrenrut replied:
I assume they are all dynamic types, because I wasn't getting any type information in pyCharm anyway. I wasn't sure what the right syntax was for these so just put them in the doc string. There are no complaints from Python when using `POINTER(VARIANT)` so I'll go with that.

leonardder commented on source/gui/accPropServer.py:
```diff
@@ -90,24 +91,53 @@ def _getPropValue(self, pIDString, dwIDStringLen, idProp):
+			return S_OK
+		if ret[1] != self.HAS_PROP:
+			return S_OK
+		pfGotProp[0] = self.HAS_PROP
```
I think I prefer

```suggestion
		pfGotProp.contents = self.HAS_PROP
```

instead. it looks clearer.

feerrenrut replied:
These libraries are very hard to read, and the documentation sucks. Though I found some help in ctypes tests, at least there are (presumably working) examples. It seems you have to actually have to do `pfGotProp.contents.value`. For a C/C++ programmer `pfGotProp[0]` probably looks more familiar when dealing with pointers.

leonardder commented on source/gui/accPropServer.py:
```diff
@@ -90,24 +91,53 @@ def _getPropValue(self, pIDString, dwIDStringLen, idProp):
+			if ret is None:
+				return S_OK
+			elif len(ret) != 2:
+				raise RuntimeError("_getPropValue implementation must return None or two element tuple")
```
I'm afraid the caller really won't like this. In this case the server and the client are the same, but I think it is better to log an error here and return a hresult that indicates an error.

feerrenrut replied:
These are both caught and logged below, perhaps this makes the control flow unclear. I'll see what I can do to make it clearer. I can't think of a situation where it is beneficial to notify the caller of an internal error, especially given the caller is something in IAccessible. . It certainly helps to log these exceptions though, otherwise they don't make it to the log.

leonardder replied:
Ah I'm sorry, I see it now. I wasn't keeping an eye on the indentation properly.
I guess it is ok if we return something else than S_OK on errors.

leonardder commented on source/gui/accPropServer.py:
```diff
@@ -90,24 +91,53 @@ def _getPropValue(self, pIDString, dwIDStringLen, idProp):
+				raise RuntimeError("_getPropValue implementation must return None or two element tuple")
+		except Exception as e:
+			log.exception()
+			return S_OK
```
Same here. Using a low level implementation now has the benefit of us changing the errors codes returned.

LeonarddeR reviewed (changes requested):
Just some small things. Having said that, I was pretty sure that I had a similar low level implementation, though yet that implementation didn't work.

Brian1Gaff commented:
Well on Windows 7, that specific error sound is gone. However I assume the 
try snap was not based on the staging snap as although restarted with add 
ons disabled it went haywire with errors on booting, which I ignored. This 
is the log of the  preferences gui.
 It all started to fall apart when I tried to change voices for a new synth, 
but then it does this on threshold anyway, a bug I think you know about, 
I'll paste the bit where I select the prefs and  all the  string of errors 
which make sounds sometimes, I have cut short.
 IO - external:inputCore.InputManager.executeGesture (08:34:28.834):
Input: kb(desktop):NVDA+n
IO - external:speech.speak (08:34:29.041):
Speaking [LangChangeCommand ('en_GB'), 'NVDA  menu']
IO - external:inputCore.InputManager.executeGesture (08:34:31.474):
Input: kb(desktop):downArrow
IO - external:speech.speak (08:34:31.495):
Speaking [LangChangeCommand ('en_GB'), 'Preferences  subMenu  p']
IO - external:inputCore.InputManager.executeGesture (08:34:33.082):
Input: kb(desktop):rightArrow
IO - external:speech.speak (08:34:33.105):
Speaking [LangChangeCommand ('en_GB'), 'Settings...  s  1 of 6']
IO - external:inputCore.InputManager.executeGesture (08:34:36.291):
Input: kb(desktop):tab
IO - external:inputCore.InputManager.executeGesture (08:34:37.219):
Input: kb(desktop):downArrow
IO - external:speech.speak (08:34:37.236):
Speaking [LangChangeCommand ('en_GB'), 'Speech dictionaries  subMenu  d']
IO - external:inputCore.InputManager.executeGesture (08:34:38.355):
Input: kb(desktop):upArrow
IO - external:speech.speak (08:34:38.368):
Speaking [LangChangeCommand ('en_GB'), 'Settings...  s  1 of 6']
IO - external:inputCore.InputManager.executeGesture (08:34:38.923):
Input: kb(desktop):enter
DEBUG - external:gui.settingsDialogs.__new__ (08:34:38.927):
Creating new settings dialog (multiInstanceAllowed:False). State of 
_instances {}
DEBUG - external:windowUtils.getWindowScalingFactor (08:34:38.928):
GetDpiForWindow failed, using GetDeviceCaps instead
DEBUG - external:windowUtils.getWindowScalingFactor (08:34:38.933):
GetDpiForWindow failed, using GetDeviceCaps instead
IO - external:speech.speak (08:34:39.015):
Speaking [LangChangeCommand ('en_GB'), 'NVDA Settings: General (normal 
configuration)  dialog']
IO - external:speech.speak (08:34:39.016):
Speaking [LangChangeCommand ('en_GB'), 'Categories:  list']
IO - external:speech.speak (08:34:39.023):
Speaking [LangChangeCommand ('en_GB'), 'General  1 of 13']
IO - external:inputCore.InputManager.executeGesture (08:34:44.315):
Input: kb(desktop):rightArrow
IO - external:inputCore.InputManager.executeGesture (08:34:44.891):
Input: kb(desktop):tab
IO - external:speech.speak (08:34:44.918):
Speaking [LangChangeCommand ('en_GB'), 'General  property page']
IO - external:speech.speak (08:34:44.925):
Speaking [LangChangeCommand ('en_GB'), 'Language (requires restart to fully 
take effect):  combo box  User default  collapsed  Alt+l']
IO - external:inputCore.InputManager.executeGesture (08:34:48.298):
Input: kb(desktop):shift+tab
IO - external:speech.speak (08:34:48.325):
Speaking [LangChangeCommand ('en_GB'), 'Categories:  list']
IO - external:speech.speak (08:34:48.333):
Speaking [LangChangeCommand ('en_GB'), 'General  1 of 13']
IO - external:inputCore.InputManager.executeGesture (08:34:49.050):
Input: kb(desktop):downArrow
DEBUG - external:windowUtils.getWindowScalingFactor (08:34:49.051):
GetDpiForWindow failed, using GetDeviceCaps instead
DEBUG - 
external:gui.settingsDialogs.SettingsDialog._setInstanceDestroyedState 
(08:34:49.053):
Setting state to destroyed for instance: 
<gui.settingsDialogs.NVDASettingsDialog object at 0x05B51DA0>
Current _instances {<gui.settingsDialogs.NVDASettingsDialog object at 
0x05B51DA0>: 0}
DEBUG - external:windowUtils.getWindowScalingFactor (08:34:49.054):
GetDpiForWindow failed, using GetDeviceCaps instead
IO - external:speech.speak (08:34:49.129):
Speaking [LangChangeCommand ('en_GB'), 'Speech  2 of 13']
IO - external:inputCore.InputManager.executeGesture (08:34:50.146):
Input: kb(desktop):tab
IO - external:speech.speak (08:34:50.202):
Speaking [LangChangeCommand ('en_GB'), 'Speech  property page']
IO - external:speech.speak (08:34:50.208):
Speaking [LangChangeCommand ('en_GB'), 'Synthesizer  grouping']
IO - external:speech.speak (08:34:50.214):
Speaking [LangChangeCommand ('en_GB'), 'Synthesizer  edit  read only  multi 
line  Alt+s']
IO - external:speech.speak (08:34:50.215):
Speaking [LangChangeCommand ('en_GB'), 'align left', 'eSpeak NG\r']
IO - external:inputCore.InputManager.executeGesture (08:34:54.971):
Input: kb(desktop):tab
IO - external:speech.speak (08:34:54.993):
Speaking [LangChangeCommand ('en_GB'), 'Change...  button  Alt+h']
IO - external:inputCore.InputManager.executeGesture (08:34:56.251):
Input: kb(desktop):space
IO - external:speech.speak (08:34:56.263):
Speaking [LangChangeCommand ('en_GB'), 'space', EndUtteranceCommand()]
IO - external:speech.speak (08:34:56.264):
Speaking [LangChangeCommand ('en_GB'), 'pressed']
DEBUG - external:gui.settingsDialogs.__new__ (08:34:56.363):
Creating new settings dialog (multiInstanceAllowed:True). State of 
_instances {<gui.settingsDialogs.NVDASettingsDialog object at 0x05B51DA0>: 
1}
DEBUG - external:windowUtils.getWindowScalingFactor (08:34:56.364):
GetDpiForWindow failed, using GetDeviceCaps instead
DEBUGWARNING - external:synthDriverHandler.getSynthList (08:34:56.370):
Synthesizer 'oneCore' doesn't pass the check, excluding from list
IO - external:speech.speak (08:34:56.417):
Speaking [LangChangeCommand ('en_GB'), 'Select Synthesizer  dialog']
IO - external:speech.speak (08:34:56.424):
Speaking [LangChangeCommand ('en_GB'), 'Synthesizer:  combo box  eSpeak NG 
collapsed  Alt+s']
IO - external:inputCore.InputManager.executeGesture (08:35:01.315):
Input: kb(desktop):downArrow
IO - external:speech.speak (08:35:01.328):
Speaking [LangChangeCommand ('en_GB'), 'Microsoft Speech API version 4']
IO - external:inputCore.InputManager.executeGesture (08:35:03.874):
Input: kb(desktop):downArrow
IO - external:speech.speak (08:35:03.885):
Speaking [LangChangeCommand ('en_GB'), 'Microsoft Speech API version 5']
IO - external:inputCore.InputManager.executeGesture (08:35:08.162):
Input: kb(desktop):tab
IO - external:speech.speak (08:35:08.182):
Speaking [LangChangeCommand ('en_GB'), 'Output device:  combo box  Speakers 
(4- BEHRINGER USB WDM   collapsed  Alt+d']
IO - external:inputCore.InputManager.executeGesture (08:35:12.675):
Input: kb(desktop):tab
IO - external:speech.speak (08:35:12.690):
Speaking [LangChangeCommand ('en_GB'), 'OK  button']
IO - external:inputCore.InputManager.executeGesture (08:35:13.787):
Input: kb(desktop):space
IO - external:speech.speak (08:35:13.804):
Speaking [LangChangeCommand ('en_GB'), 'space', EndUtteranceCommand()]
IO - external:speech.speak (08:35:13.805):
Speaking [LangChangeCommand ('en_GB'), 'pressed']
DEBUG - external:speechDictHandler.SpeechDict.load (08:35:14.833):
Loading speech dictionary 
'C:\Users\Brian\AppData\Roaming\nvda\speechDicts\voiceDicts.v1\sapi5\sapi5-ScanSoft 
Daniel_Full_22kHz.dic'...
DEBUG - external:speechDictHandler.SpeechDict.load (08:35:14.839):
1 loaded records.
DEBUG - external:synthDriverHandler.SynthDriver.loadSettings (08:35:14.839):
Loaded settings for SynthDriver sapi5
INFO - external:synthDriverHandler.setSynth (08:35:14.839):
Loaded synthDriver sapi5
DEBUG - 
external:gui.settingsDialogs.SettingsDialog._setInstanceDestroyedState 
(08:35:14.840):
Setting state to destroyed for instance: 
<gui.settingsDialogs.SynthesizerSelectionDialog object at 0x064C5F30>
Current _instances {<gui.settingsDialogs.NVDASettingsDialog object at 
0x05B51DA0>: 1, <gui.settingsDialogs.SynthesizerSelectionDialog object at 
0x064C5F30>: 0}
DEBUG - 
external:gui.settingsDialogs.SettingsDialog._setInstanceDestroyedState 
(08:35:14.841):
Setting state to destroyed for instance: 
<gui.settingsDialogs.SynthesizerSelectionDialog object at 0x064C5F30>
Current _instances {<gui.settingsDialogs.NVDASettingsDialog object at 
0x05B51DA0>: 1, <gui.settingsDialogs.SynthesizerSelectionDialog object at 
0x064C5F30>: 1}
DEBUG - 
external:gui.settingsDialogs.SettingsDialog._setInstanceDestroyedState 
(08:35:14.843):
Setting state to destroyed for instance: 
<gui.settingsDialogs.SynthesizerSelectionDialog object at 0x064C5F30>
Current _instances {<gui.settingsDialogs.NVDASettingsDialog object at 
0x05B51DA0>: 1, <gui.settingsDialogs.SynthesizerSelectionDialog object at 
0x064C5F30>: 1}
DEBUG - 
external:gui.settingsDialogs.SettingsDialog._setInstanceDestroyedState 
(08:35:14.844):
Setting state to destroyed for instance: 
<gui.settingsDialogs.SynthesizerSelectionDialog object at 0x064C5F30>
Current _instances {<gui.settingsDialogs.NVDASettingsDialog object at 
0x05B51DA0>: 1, <gui.settingsDialogs.SynthesizerSelectionDialog object at 
0x064C5F30>: 1}
DEBUG - 
external:gui.settingsDialogs.SettingsDialog._setInstanceDestroyedState 
(08:35:14.845):
Setting state to destroyed for instance: 
<gui.settingsDialogs.SynthesizerSelectionDialog object at 0x064C5F30>
Current _instances {<gui.settingsDialogs.NVDASettingsDialog object at 
0x05B51DA0>: 1, <gui.settingsDialogs.SynthesizerSelectionDialog object at 
0x064C5F30>: 1}
DEBUG - 
external:gui.settingsDialogs.SettingsDialog._setInstanceDestroyedState 
(08:35:14.845):
Setting state to destroyed for instance: 
<gui.settingsDialogs.SynthesizerSelectionDialog object at 0x064C5F30>
Current _instances {<gui.settingsDialogs.NVDASettingsDialog object at 
0x05B51DA0>: 1, <gui.settingsDialogs.SynthesizerSelectionDialog object at 
0x064C5F30>: 1}
DEBUG - 
external:gui.settingsDialogs.SettingsDialog._setInstanceDestroyedState 
(08:35:14.847):
Setting state to destroyed for instance: 
<gui.settingsDialogs.SynthesizerSelectionDialog object at 0x064C5F30>
Current _instances {<gui.settingsDialogs.NVDASettingsDialog object at 
0x05B51DA0>: 1, <gui.settingsDialogs.SynthesizerSelectionDialog object at 
0x064C5F30>: 1}
DEBUG - 
external:gui.settingsDialogs.SettingsDialog._setInstanceDestroyedState 
(08:35:14.847):
Setting state to destroyed for instance: 
<gui.settingsDialogs.SynthesizerSelectionDialog object at 0x064C5F30>
Current _instances {<gui.settingsDialogs.NVDASettingsDialog object at 
0x05B51DA0>: 1, <gui.settingsDialogs.SynthesizerSelectionDialog object at 
0x064C5F30>: 1}
DEBUG - 
external:gui.settingsDialogs.SettingsDialog._setInstanceDestroyedState 
(08:35:14.848):
Setting state to destroyed for instance: 
<gui.settingsDialogs.SynthesizerSelectionDialog object at 0x064C5F30>
Current _instances {<gui.settingsDialogs.NVDASettingsDialog object at 
0x05B51DA0>: 1, <gui.settingsDialogs.SynthesizerSelectionDialog object at 
0x064C5F30>: 1}
DEBUG - 
external:gui.settingsDialogs.SettingsDialog._setInstanceDestroyedState 
(08:35:14.861):
Setting state to destroyed for instance: 
<gui.settingsDialogs.NVDASettingsDialog object at 0x05B51DA0>
Current _instances {<gui.settingsDialogs.NVDASettingsDialog object at 
0x05B51DA0>: 1, <gui.settingsDialogs.SynthesizerSelectionDialog object at 
0x064C5F30>: 1}
DEBUG - 
external:gui.settingsDialogs.SettingsDialog._setInstanceDestroyedState 
(08:35:14.861):
Setting state to destroyed for instance: 
<gui.settingsDialogs.NVDASettingsDialog object at 0x05B51DA0>
Current _instances {<gui.settingsDialogs.NVDASettingsDialog object at 
0x05B51DA0>: 1, <gui.settingsDialogs.SynthesizerSelectionDialog object at 
0x064C5F30>: 1}
DEBUG - 
external:gui.settingsDialogs.SettingsDialog._setInstanceDestroyedState 
(08:35:14.862):
Setting state to destroyed for instance: 
<gui.settingsDialogs.NVDASettingsDialog object at 0x05B51DA0>
Current _instances {<gui.settingsDialogs.NVDASettingsDialog object at 
0x05B51DA0>: 1, <gui.settingsDialogs.SynthesizerSelectionDialog object at 
0x064C5F30>: 1}
DEBUG - 
external:gui.settingsDialogs.SettingsDialog._setInstanceDestroyedState 
(08:35:14.862):
Setting state to destroyed for instance: 
<gui.settingsDialogs.NVDASettingsDialog object at 0x05B51DA0>
Current _instances {<gui.settingsDialogs.NVDASettingsDialog object at 
0x05B51DA0>: 1, <gui.settingsDialogs.SynthesizerSelectionDialog object at 
0x064C5F30>: 1}
DEBUG - 
external:gui.settingsDialogs.SettingsDialog._setInstanceDestroyedState 
(08:35:14.862):
Setting state to destroyed for instance: 
<gui.settingsDialogs.NVDASettingsDialog object at 0x05B51DA0>
Current _instances {<gui.settingsDialogs.NVDASettingsDialog object at 
0x05B51DA0>: 1, <gui.settingsDialogs.SynthesizerSelectionDialog object at 
0x064C5F30>: 1}
DEBUG - 
external:gui.settingsDialogs.SettingsDialog._setInstanceDestroyedState 
(08:35:14.862):
Setting state to destroyed for instance: 
<gui.settingsDialogs.NVDASettingsDialog object at 0x05B51DA0>
Current _instances {<gui.settingsDialogs.NVDASettingsDialog object at 
0x05B51DA0>: 1, <gui.settingsDialogs.SynthesizerSelectionDialog object at 
0x064C5F30>: 1}
DEBUG - 
external:gui.settingsDialogs.SettingsDialog._setInstanceDestroyedState 
(08:35:14.862):
Setting state to destroyed for instance: 
<gui.settingsDialogs.NVDASettingsDialog object at 0x05B51DA0>
Current _instances {<gui.settingsDialogs.NVDASettingsDialog object at 
0x05B51DA0>: 1, <gui.settingsDialogs.SynthesizerSelectionDialog object at 
0x064C5F30>: 1}
DEBUG - 
external:gui.settingsDialogs.SettingsDialog._setInstanceDestroyedState 
(08:35:14.863):
Setting state to destroyed for instance: 
<gui.settingsDialogs.NVDASettingsDialog object at 0x05B51DA0>
Current _instances {<gui.settingsDialogs.NVDASettingsDialog object at 
0x05B51DA0>: 1, <gui.settingsDialogs.SynthesizerSelectionDialog object at 
0x064C5F30>: 1}
DEBUG - 
external:gui.settingsDialogs.SettingsDialog._setInstanceDestroyedState 
(08:35:14.863):
Setting state to destroyed for instance: 
<gui.settingsDialogs.NVDASettingsDialog object at 0x05B51DA0>
Current _instances {<gui.settingsDialogs.NVDASettingsDialog object at 
0x05B51DA0>: 1, <gui.settingsDialogs.SynthesizerSelectionDialog object at 
0x064C5F30>: 1}
DEBUG - 
external:gui.settingsDialogs.SettingsDialog._setInstanceDestroyedState 
(08:35:14.863):
Setting state to destroyed for instance: 
<gui.settingsDialogs.NVDASettingsDialog object at 0x05B51DA0>
Current _instances {<gui.settingsDialogs.NVDASettingsDialog object at 
0x05B51DA0>: 1, <gui.settingsDialogs.SynthesizerSelectionDialog object at 
0x064C5F30>: 1}
DEBUG - 
external:gui.settingsDialogs.SettingsDialog._setInstanceDestroyedState 
(08:35:14.863):
Setting state to destroyed for instance: 
<gui.settingsDialogs.NVDASettingsDialog object at 0x05B51DA0>
Current _instances {<gui.settingsDialogs.NVDASettingsDialog object at 
0x05B51DA0>: 1, <gui.settingsDialogs.SynthesizerSelectionDialog object at 
0x064C5F30>: 1}
DEBUG - 
external:gui.settingsDialogs.SettingsDialog._setInstanceDestroyedState 
(08:35:14.864):
Setting state to destroyed for instance: 
<gui.settingsDialogs.NVDASettingsDialog object at 0x05B51DA0>
Current _instances {<gui.settingsDialogs.NVDASettingsDialog object at 
0x05B51DA0>: 1, <gui.settingsDialogs.SynthesizerSelectionDialog object at 
0x064C5F30>: 1}
DEBUG - 
external:gui.settingsDialogs.SettingsDialog._setInstanceDestroyedState 
(08:35:14.864):
Setting state to destroyed for instance: 
<gui.settingsDialogs.NVDASettingsDialog object at 0x05B51DA0>
Current _instances {<gui.settingsDialogs.NVDASettingsDialog object at 
0x05B51DA0>: 1, <gui.settingsDialogs.SynthesizerSelectionDialog object at 
0x064C5F30>: 1}
DEBUG - 
external:gui.settingsDialogs.SettingsDialog._setInstanceDestroyedState 
(08:35:14.864):
Setting state to destroyed for instance: 
<gui.settingsDialogs.NVDASettingsDialog object at 0x05B51DA0>
Current _instances {<gui.settingsDialogs.NVDASettingsDialog object at 
0x05B51DA0>: 1, <gui.settingsDialogs.SynthesizerSelectionDialog object at 
0x064C5F30>: 1}
DEBUG - 
external:gui.settingsDialogs.SettingsDialog._setInstanceDestroyedState 
(08:35:14.864):
Setting state to destroyed for instance: 
<gui.settingsDialogs.NVDASettingsDialog object at 0x05B51DA0>
Current _instances {<gui.settingsDialogs.NVDASettingsDialog object at 
0x05B51DA0>: 1, <gui.settingsDialogs.SynthesizerSelectionDialog object at 
0x064C5F30>: 1}
DEBUG - 
external:gui.settingsDialogs.SettingsDialog._setInstanceDestroyedState 
(08:35:14.864):
Setting state to destroyed for instance: 
<gui.settingsDialogs.NVDASettingsDialog object at 0x05B51DA0>
Current _instances {<gui.settingsDialogs.NVDASettingsDialog object at 
0x05B51DA0>: 1, <gui.settingsDialogs.SynthesizerSelectionDialog object at 
0x064C5F30>: 1}
DEBUG - 
external:gui.settingsDialogs.SettingsDialog._setInstanceDestroyedState 
(08:35:14.864):
Setting state to destroyed for instance: 
<gui.settingsDialogs.NVDASettingsDialog object at 0x05B51DA0>
Current _instances {<gui.settingsDialogs.NVDASettingsDialog object at 
0x05B51DA0>: 1, <gui.settingsDialogs.SynthesizerSelectionDialog object at 
0x064C5F30>: 1}
DEBUG - 
external:gui.settingsDialogs.SettingsDialog._setInstanceDestroyedState 
(08:35:14.865):
Setting state to destroyed for instance: 
<gui.settingsDialogs.NVDASettingsDialog object at 0x05B51DA0>
Current _instances {<gui.settingsDialogs.NVDASettingsDialog object at 
0x05B51DA0>: 1, <gui.settingsDialogs.SynthesizerSelectionDialog object at 
0x064C5F30>: 1}
DEBUG - 
external:gui.settingsDialogs.SettingsDialog._setInstanceDestroyedState 
(08:35:14.865):
Setting state to destroyed for instance: 
<gui.settingsDialogs.NVDASettingsDialog object at 0x05B51DA0>
Current _instances {<gui.settingsDialogs.NVDASettingsDialog object at 
0x05B51DA0>: 1, <gui.settingsDialogs.SynthesizerSelectionDialog object at 
0x064C5F30>: 1}
DEBUG - 
external:gui.settingsDialogs.SettingsDialog._setInstanceDestroyedState 
(08:35:14.865):
Setting state to destroyed for instance: 
<gui.settingsDialogs.NVDASettingsDialog object at 0x05B51DA0>
Current _instances {<gui.settingsDialogs.NVDASettingsDialog object at 
0x05B51DA0>: 1, <gui.settingsDialogs.SynthesizerSelectionDialog object at 
0x064C5F30>: 1}
DEBUG - 
external:gui.settingsDialogs.SettingsDialog._setInstanceDestroyedState 
(08:35:14.865):
Setting state to destroyed for instance: 
<gui.settingsDialogs.NVDASettingsDialog object at 0x05B51DA0>
Current _instances {<gui.settingsDialogs.NVDASettingsDialog object at 
0x05B51DA0>: 1, <gui.settingsDialogs.SynthesizerSelectionDialog object at 
0x064C5F30>: 1}
DEBUG - 
external:gui.settingsDialogs.SettingsDialog._setInstanceDestroyedState 
(08:35:14.865):
Setting state to destroyed for instance: 
<gui.settingsDialogs.NVDASettingsDialog object at 0x05B51DA0>
Current _instances {<gui.settingsDialogs.NVDASettingsDialog object at 
0x05B51DA0>: 1, <gui.settingsDialogs.SynthesizerSelectionDialog object at 
0x064C5F30>: 1}
DEBUG - 
external:gui.settingsDialogs.SettingsDialog._setInstanceDestroyedState 
(08:35:14.866):
Setting state to destroyed for instance: 
<gui.settingsDialogs.NVDASettingsDialog object at 0x05B51DA0>
Current _instances {<gui.settingsDialogs.NVDASettingsDialog object at 
0x05B51DA0>: 1, <gui.settingsDialogs.SynthesizerSelectionDialog object at 
0x064C5F30>: 1}
DEBUG - 
external:gui.settingsDialogs.SettingsDialog._setInstanceDestroyedState 
(08:35:14.866):
Setting state to destroyed for instance: 
<gui.settingsDialogs.NVDASettingsDialog object at 0x05B51DA0>
Current _instances {<gui.settingsDialogs.NVDASettingsDialog object at 
0x05B51DA0>: 1, <gui.settingsDialogs.SynthesizerSelectionDialog object at 
0x064C5F30>: 1}
IO - external:speech.speak (08:35:14.919):
Speaking [LangChangeCommand ('en_GB'), 'NVDA Settings: Speech (normal 
configuration)  dialog']
DEBUGWARNING - external:synthDrivers.sapi5.SynthDriver.speak (08:35:14.920):
Unsupported speech command: LangChangeCommand ('en_GB')
IO - external:speech.speak (08:35:14.958):
Speaking [LangChangeCommand ('en_GB'), 'Speech  property page']
IO - external:speech.speak (08:35:14.964):
Speaking [LangChangeCommand ('en_GB'), 'Synthesizer  grouping']
IO - external:speech.speak (08:35:14.971):
Speaking [LangChangeCommand ('en_GB'), 'Synthesizer  edit  read only  multi 
line  Alt+s']
IO - external:speech.speak (08:35:14.972):
Speaking [LangChangeCommand ('en_GB'), 'align left', 'Microsoft Speech API 
version 5\r']
DEBUG - 
external:gui.settingsDialogs.SettingsDialog._setInstanceDestroyedState 
(08:35:14.976):
Setting state to destroyed for instance: 
<gui.settingsDialogs.SynthesizerSelectionDialog object at 0x064C5F30>
Current _instances {<gui.settingsDialogs.NVDASettingsDialog object at 
0x05B51DA0>: 1, <gui.settingsDialogs.SynthesizerSelectionDialog object at 
0x064C5F30>: 1}

It got silly around this point!

Brian

bglists@blueyonder.co.uk
Sent via blueyonder.
Please address personal E-mail to:-
briang1@blueyonder.co.uk, putting 'Brian Gaff'
in the display name field.
Newsgroup monitored: alt.comp.blind-users

feerrenrut commented:
> I was pretty sure that I had a similar low level implementation

I did't actually look at your implementation. The key things that took me time to realise:
- understanding that the variant.vt value needs to be set to VT_EMPTY before assigning a value.
- ensuring that all exceptions are caught and logged in the function so you know when there is an error.

feerrenrut committed:
Review actions for #9900

- Encompass getProp in a try block, ensure all exceptions are caught and
logged.
- Tidy type hints
- Use clearer method to set pointer values
- Remove redefined S_OK

leonardder commented on source/gui/accPropServer.py:
```diff
@@ -90,24 +103,58 @@ def _getPropValue(self, pIDString, dwIDStringLen, idProp):
-		except Exception:
+			# Preset values for "no prop value", in case we return early.
+			pfGotProp.contents.value = self.DOES_NOT_HAVE_PROP
+			pvarValue.contents.vt = VT_EMPTY
```
Here you're setting contents.vt and below you're setting contents.value. Why is that?

If you prefer subscripting with [0], I can certainly live with that. Just not sure what's the preferred way of doing this.

feerrenrut replied:
`.vt` must be set, to ensure that comtypes does not attempt to clear an uninitialised `VARIANT` , or more accurately, it's safe when it does clear it. When you assign a value, comtypes automatically converts the value being assigned, but first it clears the `VARIANT`. Out params don't seem to have a guarantee of being initialised. So `.vt` which, I assume stands for variant type, could be anything. It's explicitly unsafe to clear an uninitialised variant, see [variantClear Docs](https://docs.microsoft.com/en-us/windows/win32/api/oleauto/nf-oleauto-variantclear) for the following snippet:

> Do not use VariantClear on unitialized variants; use VariantInit to initialize a new VARIANTARG or VARIANT.

Rather than setting `.vt` we could call `VariantInit`
I suspect this was causing the issue I was having when trying to override name a while back.

I don't really have a preference for getting to the pointer contents, I'll leave it as is.

feerrenrut replied:
I'm leaning towards using VariantInit rather than setting vt, but the [server annotation sample](https://docs.microsoft.com/en-gb/windows/win32/winauto/server-annotation-sample) does it like this.

feerrenrut replied:
[This stackoverflow answer](https://stackoverflow.com/a/23542782) convinces me to use it.

leonardder replied:
Sounds reasonable.

leonardder commented on source/gui/accPropServer.py:
```diff
@@ -90,24 +103,58 @@ def _getPropValue(self, pIDString, dwIDStringLen, idProp):
+			pfGotProp.contents.value = self.HAS_PROP
+			pvarValue.contents.value = ret[0]
+		except Exception as e: # catch and log all exceptions so they are not swallowed by caller.
 			log.exception()
```
I think I still prefer returning something else than S_OK here. May be S_FALSE is a good one here, or comtypes.hresult.E_FAIL? Not sure how hresult codes work.

feerrenrut replied:
I'm not so sure, I don't think we should be reporting an error unless it is the callers fault. If we would succeed with another value, then this essentially translates to a "look-up error", that outcome is handled by the `pfGotProp` outparam.

leonardder replied:
Ah, fair enough.

feerrenrut committed:
Clarify variant init

leonardder commented:
@feerrenrut: related to this is the list of monkey patches we apply to comtypes.

From code inspection/investigation, it turns out that at least the patch to support byref in variants can be removed.
